kauri/tree.go

```diff
@@ -0,0 +1,170 @@
+package kauri
+
+import (
+	"math"
+
+	"github.com/relab/hotstuff"
+)
+
+// MaxChild determines the branching factor of the tree.
+const MaxChild = 2
+
+// TreeConfiguration is an abstraction for a tree communication model.
+type TreeConfiguration interface {
+	InitializeWithPIDs(ids map[hotstuff.ID]int)
+	GetHeight() int
+	GetChildren() []hotstuff.ID
+	GetSubTreeNodes() []hotstuff.ID
+	GetParent() (hotstuff.ID, bool)
+}
+
+// FaultFreeTree implements a fault free tree configuration.
+type FaultFreeTree struct {
+	ID                  hotstuff.ID
+	ConfigurationLength int
+	height              int
+	idToPosMapping      map[hotstuff.ID]int
+	posToIDMapping      map[int]hotstuff.ID
+}
+
+// CreateTree Creates the tree configuration, currently only fault free tree configuration is supported.
+func CreateTree(configurationLength int, myID hotstuff.ID) TreeConfiguration {
+
+	if configurationLength <= 0 {
+		return nil
+	}
+	temp := configurationLength
+	temp = temp - 1 //root
+	height := 1
+	for i := 1; temp > 0; i++ {
```

Comment:
Maybe this could be converted to a helper function:
```go
func height(maxChild float64, treeNodes int) int {
	remainingNodes := treeNodes - 1 // root
	height := 1
	for i := 1; remainingNodes > 0; i++ {
		remainingNodes -= int(math.Pow(maxChild, float64(i)))
		height++
	}
	return height
}
```
And you could add this test:
```go
func TestTreeHeight(t *testing.T) {
	const MaxChild = 2
	tests := []struct {
		configurationLength int
		expectedHeight      int
	}{
		{1, 1},
		{2, 2},
		{3, 2},
		{4, 3},
		{5, 3},
		{6, 3},
		{7, 3},
		{8, 4},
		{9, 4},
		{10, 4},
		{11, 4},
		{12, 4},
		{13, 4},
		{14, 4},
		{15, 4},
		{16, 5},
		{17, 5},
	}
	for _, test := range tests {
		actualHeight := height(MaxChild, test.configurationLength)
		if actualHeight != test.expectedHeight {
			t.Errorf("height(%d) = %d, expected %d", test.configurationLength, actualHeight, test.expectedHeight)
		}
	}
}
```